core/drivers/pm/sam/at91_pm.c

```diff
@@ -46,6 +47,30 @@ static struct at91bootstrap_bu {
 static vaddr_t at91_suspend_sram_base;
 static void (*at91_suspend_sram_fn)(struct at91_pm_data *);
 
+enum sm_handler_ret at91_pm_set_suspend_mode(struct thread_smc_args *args)
+{
+	unsigned int mode = args->a1;
+
+	if (mode > AT91_PM_BACKUP) {
+		args->a0 = SAMA5_SMC_SIP_RETURN_EINVAL;
+		return SM_HANDLER_SMC_HANDLED;
+	}
+	DMSG("Setting suspend mode to %u", mode);
+
+	args->a0 = SAMA5_SMC_SIP_RETURN_SUCCESS;
+	soc_pm.mode = mode;
```

Comment:
Does this need to be protected from concurrent accesses?
What if this is written while it's used during suspend/resume?

Reply:
The sama5d2 platform is a single core platform so this can not happen, once the `SYSTEM_SUSPEND` PSCI call has been issued, there should be no possibility to send another SMC since the CPU will be in secure world executing the previous one.
But maybe I'm missing something here.

Reply:
Yes sure !